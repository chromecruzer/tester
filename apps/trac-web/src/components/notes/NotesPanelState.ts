import {AuditNoteRecord, NullableString} from "@trac/datatypes";
import _ from "lodash";
export type NotesPanelDisplay = ('All' | 'Hide' | 'Audit' | 'Item');
export interface NotesPanelControls {
  display: NotesPanelDisplay;
  itemUuid: NullableString;
}
export default class NotesPanelState {
  public itemUuid: NullableString;
  public notes: AuditNoteRecord[];
  public display: NotesPanelDisplay;
  constructor(controls: NotesPanelControls) {
    this.notes = [];
    this.itemUuid = null;
    this.display = 'Hide'
    this.setDisplay(controls.display);
    this.selectItemNote(controls.itemUuid);
  }
  public setNotes(newNotes) {
    this.notes = [...newNotes];
  }
  public setDisplay(display) {
    // console.log(`Setting display to ${display}`)
    switch(display) {
      case 'Audit':
        this.display = this.hasAuditNotes() ? display : 'Hide';
        break;
      default:
        this.display = display;
    }
  }
  public selectItemNote(uuid = null as NullableString) {
    if(uuid) {
      this.display = 'Item';
    }
    this.itemUuid = uuid;
  }
  public setControls(controls: NotesPanelControls) {
  }
  public getControls() {
    return { display: this.display, itemUuid: this.itemUuid};
  }
  public hasAuditNotes() {
    return _.filter(this.notes, n => n.annotate_type === 'Audit').length > 0;
  }
  public getFilteredNotes() {
    switch(this.display) {
      case 'Item':
        if(this.itemUuid) {
          // console.log(`item filtered notes ${this.itemUuid} and display ${this.display}`);
          return _.filter(this.notes, n => n.annotate_type !== 'Audit' && n.annotated_uuid === this.itemUuid);
        }
        return [];
      case 'All':
        // console.log(`no filtering display ${this.display}`);
        return this.notes;
      case 'Audit':
        // console.log(`audit filtered notes display ${this.display}`);
        return _.filter(this.notes, n => n.annotate_type === 'Audit');
      case "Hide":
        // console.log(`hide notes display ${this.display}`);
        return [];
    }
  }
}
