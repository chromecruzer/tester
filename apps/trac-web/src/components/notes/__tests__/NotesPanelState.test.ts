import NotesPanelState from "../NotesPanelState";
import _ from "lodash";

describe('NotesPanelState', () => {
  let notes, state;
  beforeEach(() => {
    notes = [
      {
        annotate_type: 'Audit',
        annotated_uuid: 'audit1'
      },
      {
        annotate_type: 'Audit',
        annotated_uuid: 'audit2'
      },
      {
        annotate_type: 'Item',
        annotated_uuid: 'item'
      },
      {
        annotate_type: 'Match',
        annotated_uuid: 'item'
      },
      {
        annotate_type: 'Item',
        annotated_uuid: 'other item'
      },
    ];
    state = new NotesPanelState();
  });
  describe('setNotes', () => {
    it('should set the notes', () => {
      state.setNotes(notes);
      expect(state.notes).toEqual(notes);
    });
  });
  describe('setDisplay', () => {
    it('should set the display state correctly', () => {
      expect(state.display).toEqual('Hide');
      state.setDisplay('All');
      expect(state.display).toEqual('All');
    });
  });
  describe('itemFilter', () => {
    it('should set uuid', () => {
      expect(state.itemUuid).toBeNull();
      state.itemFilter('item');
      expect(state.itemUuid).toEqual('item');
    });
    it('should set the display to Item when there is a uuid', () => {
      expect(state.itemUuid).toBeNull();
      state.itemFilter('item');
      expect(state.itemUuid).toEqual('item');
      expect(state.display).toEqual('Item');
      state.setNotes(notes);
      state.setDisplay('Audit');
      state.itemFilter(null);
      expect(state.itemUuid).toBeNull();
      expect(state.display).toEqual('Audit');
    });
  it('should set the display not display audit notes if their are none', () => {
    state.setNotes(_.filter(notes, n => n.annotate_type !== 'Audit'));
    state.setDisplay('Audit');
    state.itemFilter(null);
    expect(state.itemUuid).toBeNull();
    expect(state.display).toEqual('Hide');
  });
});
  describe('getNotes', () => {
    it('should filter notes by item', () => {
      state.setNotes(notes);
      state.itemFilter('item');
      const actual = state.getNotes();
      expect(actual).toMatchSnapshot();
    });
    it('should return all notes if specified', () => {
      state.setNotes(notes);
      state.setDisplay('All');
      const actual = state.getNotes();
      expect(actual).toMatchSnapshot();
    });
    it('should return only audit notes', () => {
      state.setNotes(notes);
      state.setDisplay('Audit');
      const actual = state.getNotes();
      expect(actual).toMatchSnapshot();
    });
    it('should return an empty array when hidden', () => {
      state.setNotes(notes);
      const actual = state.getNotes();
      expect(actual).toEqual([]);
    });
  });
});
