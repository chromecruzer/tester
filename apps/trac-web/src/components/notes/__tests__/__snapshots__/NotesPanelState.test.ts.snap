// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotesPanelState getNotes should filter notes by item 1`] = `
Array [
  Object {
    "annotate_type": "Item",
    "annotated_uuid": "item",
  },
  Object {
    "annotate_type": "Match",
    "annotated_uuid": "item",
  },
]
`;

exports[`NotesPanelState getNotes should return all notes if specified 1`] = `
Array [
  Object {
    "annotate_type": "Audit",
    "annotated_uuid": "audit1",
  },
  Object {
    "annotate_type": "Audit",
    "annotated_uuid": "audit2",
  },
  Object {
    "annotate_type": "Item",
    "annotated_uuid": "item",
  },
  Object {
    "annotate_type": "Match",
    "annotated_uuid": "item",
  },
  Object {
    "annotate_type": "Item",
    "annotated_uuid": "other item",
  },
]
`;

exports[`NotesPanelState getNotes should return only audit notes 1`] = `
Array [
  Object {
    "annotate_type": "Audit",
    "annotated_uuid": "audit1",
  },
  Object {
    "annotate_type": "Audit",
    "annotated_uuid": "audit2",
  },
]
`;
